tls-client/src/handshaker.rs

```diff
@@ -1,39 +1,234 @@
+        let p1 = hmac_sha256(&pms, &a1_seed);
+        let p2 = hmac_sha256(&pms, &a2_seed);
+        let mut ms = [0u8; 48];
+        ms[..32].copy_from_slice(&p1);
+        ms[32..].copy_from_slice(&p2[..16]);
+        // store ms (for CF, SF later)
+        self.master_secret = Some(ms);
+
+        let seed = seed_ke(&client_random, &server_random);
+        let a1 = hmac_sha256(&ms, &seed);
+        let a2 = hmac_sha256(&ms, &a1);
+        let mut a1_seed = [0u8; 109];
+        a1_seed[..32].copy_from_slice(&a1);
+        a1_seed[32..].copy_from_slice(&seed);
+        let mut a2_seed = [0u8; 109];
+        a2_seed[..32].copy_from_slice(&a2);
+        a2_seed[32..].copy_from_slice(&seed);
+        let p1 = hmac_sha256(&ms, &a1_seed);
+        let p2 = hmac_sha256(&ms, &a2_seed);
+        let mut ek = [0u8; 40];
+        ek[..32].copy_from_slice(&p1);
+        ek[32..].copy_from_slice(&p2[..8]);
+        // store session keys
+        self.expanded_keys = Some(ek);
+
+        Ok(())
     }
-    async fn set_server_key_share(&mut self, _key: PublicKey) -> Result<(), Self::Error> {
-        Err(Error::General("handshaker not yet available".to_string()))
+
+    async fn client_key_share(&mut self) -> Result<PublicKey, Error> {
+        // return our ecdh pubkey
+        Ok(PublicKey {
+            group: NamedGroup::secp256r1,
+            key: self.ecdh_pubkey.unwrap(),
+        })
     }
+
     async fn set_hs_hash_server_hello(&mut self, _hash: &[u8]) -> Result<(), Self::Error> {
-        Err(Error::General("handshaker not yet available".to_string()))
+        Ok(())
+        // will be used only in 2PC
+
+        // TODO the handshake hash is not up to Server Hello but must
+        // be up to Client Key Exchange, so this fn should be called
+        // receive_hs_hash_client_key_exchange
     }
-    async fn server_finished(&mut self, _hash: &[u8]) -> Result<Vec<u8>, Self::Error> {
-        Err(Error::General("handshaker not yet available".to_string()))
+
+    async fn client_finished(&mut self, hs_hash: &[u8]) -> Result<Vec<u8>, Self::Error> {
+        // TODO we must also return plaintext CF so that tls-client could update
+        // the handshake hash
+
+        // compute CF from ms + hash
+        // PRF as in test_prf() in tls-core/src/handshake/mod.rs
+        let ms = self.master_secret.unwrap();
+        let mut seed = [0u8; 47];
+        seed[..15].copy_from_slice(b"client finished");
+        seed[15..].copy_from_slice(hs_hash);
+        let a1 = hmac_sha256(&ms, &seed);
+        let mut a1_seed = [0u8; 79];
+        a1_seed[..32].copy_from_slice(&a1);
+        a1_seed[32..].copy_from_slice(&seed);
+        let mut verify_data = [0u8; 12];
+        verify_data.copy_from_slice(&hmac_sha256(&ms, &a1_seed)[..12]);
+        let mut client_finished = [0u8; 16];
+        client_finished[..4].copy_from_slice(&[0x14, 0x00, 0x00, 0x0C]);
+        client_finished[4..].copy_from_slice(&verify_data);
+
+        // encrypt client finished
```

Comment:
This function does not encrypt the ClientFinished, the return value should simply be the verify_data bytes. The tls-client will construct the plaintext CF and pass it to encrypt()